examples/water_sampling_hrex.py

```diff
@@ -0,0 +1,355 @@
+
+    ligand_idxs = np.arange(num_water_atoms, num_water_atoms + mol.GetNumAtoms())
+
+    # re-parameterize ligand idxs to do decouple only half of the buckyball
+    nb_params = params[-1]
+
+    final_masses = combined_masses
+    final_conf = np.concatenate([solvent_conf, get_romol_conf(mol)], axis=0)
+
+    component_dim = 4  # q,s,e,w
+    num_atoms = len(final_conf)
+
+    ligand_water_flat_idxs = np.s_[component_dim * num_atoms : 2 * component_dim * num_atoms]  # water
+    ligand_protein_flat_idxs = np.s_[2 * component_dim * num_atoms : 3 * component_dim * num_atoms :]  # non-water
+    ligand_water_params = nb_params[ligand_water_flat_idxs].reshape(-1, 4)
+    ligand_protein_params = nb_params[ligand_protein_flat_idxs].reshape(-1, 4)
+
+    fully_coupled_ligand_atoms = ligand_idxs[: len(ligand_idxs) // 2]
+
+    ligand_water_params = ligand_water_params.at[fully_coupled_ligand_atoms, -1].set(0)
+    ligand_protein_params = ligand_protein_params.at[fully_coupled_ligand_atoms, -1].set(0)
+
+    nb_params = nb_params.at[ligand_water_flat_idxs].set(ligand_water_params.reshape(-1))
+    nb_params = nb_params.at[ligand_protein_flat_idxs].set(ligand_protein_params.reshape(-1))
+
+    # override last potential with the updated nb_params
+    host_bps = []
+    for p, bp in zip(params[:-1], potentials[:-1]):
+        host_bps.append(bp.bind(p))
+
+    host_bps.append(potentials[-1].bind(nb_params))
+
+    # (YTZ): This is disabled because the initial ligand and starting waters are pre-minimized
+    # print("Minimizing the host system...", end="", flush=True)
+    # host_conf = minimize_host_4d([mol], host_config, ff)
+    # print("Done", flush=True)
+
+    temperature = DEFAULT_TEMP
+    dt = 1e-3
+    barostat_interval = 25
+    integrator = LangevinIntegrator(temperature, dt, 1.0, final_masses, seed)
+    bond_list = get_bond_list(host_bps[0].potential)
+    group_idxs = get_group_indices(bond_list, len(final_masses))
+    barostat = MonteCarloBarostat(
+        len(final_masses), DEFAULT_PRESSURE, temperature, group_idxs, barostat_interval, seed + 1
+    )
+
+    initial_state = InitialState(
+        host_bps, integrator, barostat, final_conf, np.zeros_like(final_conf), solvent_box, lamb, ligand_idxs
+    )
+
+    num_water_mols = num_water_atoms // 3
+
+    return initial_state, num_water_mols, solvent_topology
+
+
+def compute_density(n_waters, box):
+    box_vol = np.prod(np.diag(box))
+    numerator = n_waters * 18.01528 * 1e27
+    denominator = box_vol * AVOGADRO * 1000
+    return numerator / denominator
+
+
+def compute_occupancy(x_t, box_t, ligand_idxs, threshold):
+    # compute the number of waters inside the buckyball ligand
+
+    num_ligand_atoms = len(ligand_idxs)
+    num_host_atoms = len(x_t) - num_ligand_atoms
+    host_coords = x_t[:num_host_atoms]
+    ligand_coords = x_t[num_host_atoms:]
+    ligand_centroid = np.mean(ligand_coords, axis=0)
+    count = 0
+    for rj in host_coords:
```

Comment:
nit: could this be vectorized with a small change? E.g. something like

```python
diffs = exchange_mover.delta_r_np(ligand_centroid[:, None], host_coords, box_t)
dijs = np.linalg.norm(diffs, axis=0)
count = np.sum(diffs < threshold)
```

Reply:
will try

Reply:
had to swap `[:, None]` with `[None, :]` but otherwise works
```
diffs = exchange_mover.delta_r_np(ligand_centroid[None, :], host_coords, box_t)
dijs = np.linalg.norm(diffs)
count = np.sum(dijs < threshold)
```

Reply:
(also `np.linalg.norm` is over 1, not 0)